File: fabtrk/fabtrk/report/ft_drawing_report/ft_drawing_report.js
// Copyright (c) 2026, UpGo Technologies and contributors
// For license information, please see license.txt

frappe.query_reports["FT Drawing Report"] = {
	onload(report) {

		frappe.query_report.set_filter_value("project_number", []);
		frappe.query_report.set_filter_value("drawing_number", []);
		frappe.query_report.set_filter_value("is_active", 1);

		// Excel Button
		report.page.add_inner_button("Download Excel", function () {

			let filters = report.get_values();

			let url =
				"/api/method/fabtrk.fabtrk.report.ft_drawing_report.ft_drawing_report.download_drawing_excel"
				+ "?filters=" + encodeURIComponent(JSON.stringify(filters));

			window.location.href = url;

		});

		setTimeout(() => {
			frappe.query_report.refresh();
		}, 100);
	},

	after_datatable_render(datatable) {
		render_custom_summary();
	},

	filters: [

		// ---------------- PROJECT ----------------
		{
			fieldname: "project_number",
			label: "Project Number",
			fieldtype: "MultiSelectList",

			get_data: function (txt) {

				let is_active = frappe.query_report.get_filter_value("is_active");

				let filters = {};

				if (is_active) {
					filters.is_active = 1;
				}

				return frappe.call({
					method: "frappe.client.get_list",
					args: {
						doctype: "FT Project",
						filters: filters,
						fields: ["name", "project_name", "description"]
					}
				}).then(r => {

					return (r.message || []).map(d => {
						return {
							value: d.name,
							// label: d.name + (d.project_name ? " - " + d.project_name : ""),
							label: d.name,
							description: d.description || ""
						};
					});

				});
			},

			on_change() {
				frappe.query_report.set_filter_value("drawing_number", []);
				frappe.query_report.refresh();
				setTimeout(() => {
					frappe.query_report.refresh();
				}, 100);
			}
		},

		// ---------------- DRAWING ----------------
		{
			fieldname: "drawing_number",
			label: "Drawing Number",
			fieldtype: "MultiSelectList",

			get_data: function (txt) {

				let projects = frappe.query_report.get_filter_value("project_number");
				let filters = {};

				if (projects && projects.length > 0) {
					filters.project_number = ["in", projects];
				}

				return frappe.call({
					method: "frappe.client.get_list",
					args: {
						doctype: "FT Add Drawing",
						filters: filters,
						fields: ["name", "drawing_number"]
					}
				}).then(r => {

					let unique = {};
					let result = [];

					(r.message || []).forEach(d => {

						if (!unique[d.drawing_number]) {

							unique[d.drawing_number] = true;

							result.push({
								value: d.name,
								label: d.drawing_number,
								description: ""
							});

						}

					});

					return result;
				});
			},

			on_change() {
				frappe.query_report.refresh();
			}
		},

		// ---------------- ACTIVE ----------------
		{
			fieldname: "is_active",
			label: "Is Active",
			fieldtype: "Check",
			default: 1,

			on_change() {

				frappe.query_report.set_filter_value("project_number", []);
				frappe.query_report.set_filter_value("drawing_number", []);

				frappe.query_report.refresh();
			}
		}

	],

	formatter: function (value, row, column, data, default_formatter) {

		value = default_formatter(value, row, column, data);

		if (data && data.project_number === "TOTAL") {

			if (column.id === "_index") {
				value = "";
			}

			value = `<span style="font-weight:bold;background:#f2f2f2">${value}</span>`;
		}

		return value;
	}
};

// ---------------- CUSTOM SUMMARY CARDS ----------------
function render_custom_summary() {

	// Remove old summary if exists
	$(".ft-custom-summary").remove();

	let report_data = frappe.query_report.data || [];

	if (!report_data || report_data.length === 0) return;

	// ---------------- CALCULATIONS ----------------
	let data_rows = report_data.filter(d => d.project_number !== "TOTAL");

	// Total Projects
	let total_projects = new Set(data_rows.map(d => d.project_number).filter(Boolean)).size;

	// Total No of Line Items (rows with po_serial_no)
	let total_lines = data_rows.filter(d => d.po_serial_no).length;

	// Total Unique Drawings
	let total_drawings = new Set(data_rows.map(d => d.drawing_number).filter(Boolean)).size;

	// Total Weight as per Project (sum of all total_weight per project)
	let project_weight_map = {};
	data_rows.forEach(d => {
		if (d.project_number) {
			project_weight_map[d.project_number] = (project_weight_map[d.project_number] || 0) + (d.total_weight || 0);
		}
	});
	let total_weight_per_project = Object.values(project_weight_map).reduce((a, b) => a + b, 0);

	// Total Weight of Line Items Assign to Project (rows with po_serial_no)
	let total_weight_line_items = data_rows
		.filter(d => d.po_serial_no)
		.reduce((sum, d) => sum + (d.total_weight || 0), 0);

	// Drawing Balance for Customer
	let drawing_balance = total_weight_per_project - total_weight_line_items;

	// ---------------- FORMAT NUMBER ----------------
	function fmt(num) {
		return num.toLocaleString("en-IN", { minimumFractionDigits: 3, maximumFractionDigits: 3 });
	}

	// ---------------- CARDS HTML ----------------
	let html = `
	<div class="ft-custom-summary">
		<div class="ft-card">
			<p class="ft-card-label">Total Projects</p>
			<span class="ft-card-value">${total_projects}</span>
			<p class="ft-card-label" style="margin-top:12px;">Total Weight as per Project(Kg)</p>
			<span class="ft-card-value">${fmt(total_weight_per_project)}</span>
		</div>
		<div class="ft-card">
			<p class="ft-card-label">Project Total No of Line Items</p>
			<span class="ft-card-value">${total_lines}</span>
			<p class="ft-card-label" style="margin-top:12px;">Total Weight of Line Items Assign to Project(Kg)</p>
			<span class="ft-card-value">${fmt(total_weight_line_items)}</span>
		</div>
		<div class="ft-card">
			<p class="ft-card-label">Total No of Unique Drawings</p>
			<span class="ft-card-value">${total_drawings}</span>
		</div>
		<div class="ft-card">
			<p class="ft-card-label">Drawing Balance for Customer(Kg)</p>
			<span class="ft-card-value">${fmt(drawing_balance)}</span>
		</div>
	</div>`;

	// Inject above the datatable
	$(".report-wrapper").prepend(html);
}


// ---------------- STYLE ----------------

$(`<style>

.ft-custom-summary {
	display: grid;
	grid-template-columns: repeat(4, 1fr);
	gap: 20px;
	width: 100%;
	padding: 20px 0px;
}
@media (max-width: 768px) {
    .ft-custom-summary {
        grid-template-columns: repeat(1, 1fr) !important;
    }

    /* Scroll horizontal for table wrapper on small screens */
    .datatable .dt-scrollable {
        overflow-x: auto !important;
        -webkit-overflow-scrolling: touch; /* smooth scrolling on iOS */
    }
}

.ft-card {
	background: #fff;
	padding: 20px;
	border-radius: 10px;
	text-align: center;
	border: 2px solid #eef0f4;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
}

.ft-card-label {
	font-size: 14px;
	color: #555;
	margin: 0 0 4px 0;
	// text-align: center;
	// width: 100%;
}

.ft-card-value {
	font-size: 16px;
	font-weight: 600;
	color: #000;
	display: block;
	// text-align: center;
	// width: 100%;
}

.datatable .dt-cell__content {
	text-align: center !important;
	justify-content: center !important;
}

.datatable .dt-cell--col-0 .dt-cell__content {
	display: flex;
	justify-content: center;
	align-items: center;
	width: 100%;
}

.summary-container{
	display:grid;
	grid-template-columns:repeat(4,1fr);
	gap:20px;
	width:100%;
}

.summary-section{
	background:#fff;
	padding:20px;
	border-radius:10px;
	text-align:center;
	border:2px solid #eef0f4;
}

.section-content-count p{
	font-size:14px;
	color:#555;
}

.section-content-count span{
	font-size:16px;
	font-weight:600;
}

</style>`).appendTo("head");

File: fabtrk/fabtrk/report/ft_drawing_report/ft_drawing_report.py
# # Copyright (c) 2026, UpGo Technologies and contributors
# # For license information, please see license.txt

# # import frappe


# def execute(filters=None):
# 	columns, data = [], []
# 	return columns, data



import frappe
from frappe.utils import fmt_money

def execute(filters=None):

    filters = filters or {}

    columns = [
        {"label": "Project", "fieldname": "project_number", "fieldtype": "Link", "options": "FT Project", "width": 200},
        {"label": "Drawing Number", "fieldname": "drawing_number", "width": 200},
        {"label": "PO Serial No", "fieldname": "po_serial_no", "width": 190},
        {"label": "Unit Weight", "fieldname": "unit_weight", "fieldtype": "Float", "width": 200},
        {"label": "Required Qty", "fieldname": "required_qty", "fieldtype": "Float", "width": 190},
        {"label": "Total Weight", "fieldname": "total_weight", "fieldtype": "Float", "width": 200},
    ]

    conditions = ""
    values = {}

    if filters.get("project_number"):
        conditions += " AND p.name IN %(project_number)s"
        values["project_number"] = tuple(filters.get("project_number"))

    if filters.get("drawing_number"):
        conditions += " AND pod.drawing_number IN %(drawing_number)s"
        values["drawing_number"] = tuple(filters.get("drawing_number"))

    if filters.get("is_active"):
        conditions += " AND p.is_active = 1"

    data = frappe.db.sql(f"""
        SELECT
            p.name as project_number,
            ad.drawing_number,
            pod.po_serial_no,
            IFNULL(pod.unit_weight,0) as unit_weight,
            IFNULL(pod.required_qty,0) as required_qty,
            IFNULL(pod.total_weight,0) as total_weight
        FROM `tabFT Project` p
        LEFT JOIN `tabFT Po Drawing` pod
            ON p.name = pod.project_number
        LEFT JOIN `tabFT Add Drawing` ad
            ON ad.name = pod.drawing_number
        WHERE 1=1 {conditions}
        ORDER BY p.name, CAST(pod.po_serial_no AS UNSIGNED)
    """, values, as_dict=True)

    total_unit_weight = sum(d.get("unit_weight", 0) for d in data)
    total_qty         = sum(d.get("required_qty", 0) for d in data)
    total_weight      = sum(d.get("total_weight", 0) for d in data)

    data.append({
        "project_number": "TOTAL",
        "drawing_number": "",
        "po_serial_no":   "",
        "unit_weight":    total_unit_weight,
        "required_qty":   total_qty,
        "total_weight":   total_weight
    })

    return columns, data, None, None, None


@frappe.whitelist()
def download_drawing_excel(filters):

	import frappe
	import openpyxl
	from openpyxl.styles import Font, PatternFill, Border, Side, Alignment
	from io import BytesIO

	filters = frappe.parse_json(filters)

	conditions = ""
	values = {}

	if filters.get("project_number"):
		conditions += " AND p.name IN %(project_number)s"
		values["project_number"] = tuple(filters.get("project_number"))

	if filters.get("drawing_number"):
		conditions += " AND pod.drawing_number IN %(drawing_number)s"
		values["drawing_number"] = tuple(filters.get("drawing_number"))

	data = frappe.db.sql(f"""
		SELECT
			p.name as project,
			IFNULL(pod.drawing_number,'') as drawing_number,
			IFNULL(pod.po_serial_no,'') as po_serial_no,
			IFNULL(pod.unit_weight,0) as unit_weight,
			IFNULL(pod.required_qty,0) as required_qty,
			IFNULL(pod.total_weight,0) as total_weight
		FROM `tabFT Project` p
		LEFT JOIN `tabFT Po Drawing` pod 
			ON p.name = pod.project_number
		WHERE 1=1 {conditions}
		ORDER BY p.name
	""", values, as_dict=True)

	# ---------------- WORKBOOK ----------------
	wb = openpyxl.Workbook()
	ws = wb.active
	ws.title = "PO Drawing Data"

	# ---------------- HEADING ----------------
	ws.merge_cells('A1:F1')

	title_cell = ws['A1']
	title_cell.value = "PO Drawing"
	title_cell.font = Font(size=16, bold=True)
	title_cell.alignment = Alignment(horizontal="center", vertical="center")

	# ---------------- HEADERS ----------------
	headers = [
		"Project",
		"Drawing Number",
		"PO Serial No",
		"Unit Weight",
		"Required Qty",
		"Total Weight"
	]

	header_font = Font(bold=True, size=13, color="FFFFFF")
	header_fill = PatternFill(start_color="2F75B5", end_color="2F75B5", fill_type="solid")

	data_font = Font(size=12)
	total_font = Font(bold=True, size=13, color="FFFFFF")
	total_fill = PatternFill(start_color="2F75B5", end_color="2F75B5", fill_type="solid")

	border = Border(
		left=Side(style="thin"),
		right=Side(style="thin"),
		top=Side(style="thin"),
		bottom=Side(style="thin")
	)

	center = Alignment(horizontal="center", vertical="center")

	# Header Row
	for col, header in enumerate(headers, 1):
		cell = ws.cell(row=2, column=col, value=header)
		cell.font = header_font
		cell.fill = header_fill
		cell.border = border
		cell.alignment = center

	row_no = 3

	unit_weight_total = 0
	required_qty_total = 0
	grand_total = 0

	# ---------------- DATA ----------------
	for d in data:

		unit_weight = d.get("unit_weight") or 0
		required_qty = d.get("required_qty") or 0
		total_weight = d.get("total_weight") or 0

		unit_weight_total += unit_weight
		required_qty_total += required_qty
		grand_total += total_weight

		row_values = [
			d.get("project"),
			d.get("drawing_number"),
			d.get("po_serial_no"),
			unit_weight,
			required_qty,
			total_weight
		]

		for col, val in enumerate(row_values, 1):

			cell = ws.cell(row=row_no, column=col, value=val)
			cell.font = data_font
			cell.border = border
			cell.alignment = center

			if col in [4,6]:
				cell.number_format = '0.000'

		row_no += 1

	# ---------------- TOTAL ROW ----------------
	total_label = ws.cell(row=row_no, column=1, value="Total")
	total_label.font = total_font
	total_label.fill = total_fill
	total_label.alignment = Alignment(horizontal="left", vertical="center")

	ws.cell(row=row_no, column=2).fill = total_fill
	ws.cell(row=row_no, column=3).fill = total_fill

	unit_total_cell = ws.cell(row=row_no, column=4, value=unit_weight_total)
	unit_total_cell.font = total_font
	unit_total_cell.fill = total_fill
	unit_total_cell.number_format = '0.000'
	unit_total_cell.alignment = center

	qty_total_cell = ws.cell(row=row_no, column=5, value=required_qty_total)
	qty_total_cell.font = total_font
	qty_total_cell.fill = total_fill
	qty_total_cell.alignment = center

	total_cell = ws.cell(row=row_no, column=6, value=grand_total)
	total_cell.font = total_font
	total_cell.fill = total_fill
	total_cell.number_format = '0.000'
	total_cell.alignment = center

	for col in range(1, 7):
		ws.cell(row=row_no, column=col).border = border

	# ---------------- COLUMN WIDTH ----------------
	ws.column_dimensions["A"].width = 20
	ws.column_dimensions["B"].width = 22
	ws.column_dimensions["C"].width = 18
	ws.column_dimensions["D"].width = 15
	ws.column_dimensions["E"].width = 15
	ws.column_dimensions["F"].width = 18

	# ---------------- FILE EXPORT ----------------
	file_stream = BytesIO()
	wb.save(file_stream)
	file_stream.seek(0)

	frappe.response['filename'] = "Drawing_Report.xlsx"
	frappe.response['filecontent'] = file_stream.getvalue()
	frappe.response['type'] = 'binary'

   